internal/publish/metrics.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package publish
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/google/exposure-notifications-server/internal/jsonutil"
+	"github.com/google/exposure-notifications-server/internal/maintenance"
+	"github.com/google/exposure-notifications-server/internal/publish/model"
+	verifyapi "github.com/google/exposure-notifications-server/pkg/api/v1"
+	"github.com/google/exposure-notifications-server/pkg/logging"
+	"go.opencensus.io/trace"
+
+	"github.com/mikehelmick/go-chaff"
+)
+
+func (h *PublishHandler) HandleMetrics() http.Handler {
+	mResponder := maintenance.New(h.config)
+	return h.tracker.HandleTrack(chaff.HeaderDetector("X-Chaff"),
+		mResponder.Handle(
+			http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				ctx, span := trace.StartSpan(r.Context(), "(*publish.HandleMetrics)")
+				defer span.End()
+
+				var request verifyapi.MetricsRequest
+				code, err := jsonutil.Unmarshal(w, r, &request)
+				if err != nil {
+					message := fmt.Sprintf("error unmarshalling API call, code: %v: %v", code, err)
+					span.SetStatus(trace.Status{Code: trace.StatusCodeInternal, Message: message})
+					errorCode := verifyapi.ErrorBadRequest
+					if code == http.StatusInternalServerError {
+						errorCode = verifyapi.ErrorInternalError
+					}
+					w.Header().Set("Content-Type", "application/json")
+					w.WriteHeader(http.StatusBadRequest)
+					fmt.Fprintf(w, "{\"error\": \"%v\", \"code\": \"%v\"}", err.Error(), errorCode)
+					return
+				}
+
+				response, status := h.handleMetricsRequest(ctx, r.Header.Get("Authorization"), &request)
+
+				data, err := json.Marshal(response)
+				if err != nil {
+					w.Header().Set("Content-Type", "application/json")
+					w.WriteHeader(http.StatusInternalServerError)
+					fmt.Fprintf(w, "{\"error\": \"%v\"}", err.Error())
+					return
+				}
+
+				w.Header().Set("Content-Type", "application/json")
```

Comment:
The content-type is always appjson - you could probably set that once higher up in the body